PayloadPkg/FirmwareUpdate/FirmwareUpdateHelper.c

```diff
@@ -18,7 +18,193 @@
+  @retval others                  Device error, the command aborts abnormally.
+
+**/
+EFI_STATUS
+EFIAPI
+BootMediaRead (
+  IN     UINT64   Address,
+  IN     UINT32   ByteCount,
+  OUT    UINT8    *Buffer
+  )
+{
+  return mFwuSpiService->SpiRead (FlashRegionBios, (UINT32)Address, ByteCount, Buffer);
+}
+
+/**
+  This function writes blocks from the SPI device.
+
+  @param[in]   Address            The block address in the FlashRegionAll to read from on the SPI.
+  @param[in]   ByteCount          Size of the Buffer in bytes.
+  @param[out]  Buffer             Pointer to the data to write.
+
+  @retval EFI_SUCCESS             Write completes successfully.
+  @retval others                  Device error, the command aborts abnormally.
+
+**/
+EFI_STATUS
+EFIAPI
+BootMediaWrite (
+  IN     UINT64   Address,
+  IN     UINT32   ByteCount,
+  OUT    UINT8    *Buffer
+  )
+{
+  return mFwuSpiService->SpiWrite (FlashRegionBios, (UINT32)Address, ByteCount, Buffer);
+}
+
+/**
+  This function erases blocks from the SPI device.
+
+  @param[in]  Address             The block address in the FlashRegionAll to read from on the SPI.
+  @param[in]  ByteCount           Size of the region to erase in bytes.
+
+  @retval EFI_SUCCESS             Erase completes successfully.
+  @retval others                  Device error, the command aborts abnormally.
+
+**/
+EFI_STATUS
+EFIAPI
+BootMediaErase (
+  IN     UINT64   Address,
+  IN     UINT32   ByteCount
+  )
+{
+  return mFwuSpiService->SpiErase (FlashRegionBios, (UINT32)Address, ByteCount);
+}
+
+/**
+  Get SVN from existing firmware
+
+  This routine get SPI base address and read first four bytes
+  to get STAGE1A FV base address, using this base address
+  this routine can calculate the offset to the SVN structure.
+
+  @param[in]  Stage1AFvPointer    Pointer to Stage1A fv base.
+  @param[out] BlVersion           Pointer to SBL version struct.
+
+  @retval EFI_SUCCESS             Read SVN success
+  @retval EFI_INVALID_PARAMETER   Invalid parameter
+**/
+EFI_STATUS
+EFIAPI
+GetSvn (
+  IN  UINT32                Stage1AFvPointer,
+  OUT BOOT_LOADER_VERSION   **BlVersion
+  )
+{
+  UINT32                Stage1AFvBase;
+  UINT32                TopSwapRegionSize;
+  EFI_STATUS            Status;
+
+  //
+  // When updating BP1 - read version from backup partition
+  // When updatint BP0 - since top swap bit is set, we need
+  //                     to substract the top swap region size
+  //                     because when top swap bit is set
+  //                     address lines will be inverted and point
+  //                     to stage 1A from primary partition.
+  //
+  GetRegionInfo(&TopSwapRegionSize, NULL, NULL);
+  Stage1AFvPointer = Stage1AFvPointer - TopSwapRegionSize;
+
+  Stage1AFvBase = (UINT32)(*(UINT32 *)(UINTN)Stage1AFvPointer);
+  Stage1AFvBase = Stage1AFvPointer - (~Stage1AFvBase + 1) + sizeof(UINT32);
+
+  Status = GetVersionfromFv (&Stage1AFvBase, FALSE, BlVersion);

```

Comment:
Stated above.  Since we don't change value for Stage1AFvBase, you should just pass the actual base.

Reply:
changed code to just pass the base